src/sources/internal_metrics.rs

```diff
@@ -0,0 +1,178 @@
+use crate::{
+    event::metric::{Metric, MetricKind, MetricValue},
+    shutdown::ShutdownSignal,
+    topology::config::{DataType, GlobalOptions, SourceConfig, SourceDescription},
+    Event,
+};
+use chrono::Utc;
+use futures::{
+    compat::Future01CompatExt,
+    future::{FutureExt, TryFutureExt},
+    stream::StreamExt,
+};
+use futures01::{sync::mpsc, Future, Sink};
+use metrics_core::Key;
+use metrics_runtime::{Controller, Measurement};
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, time::Duration};
+use tokio::time::interval;
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+pub struct InternalMetricsConfig;
+
+inventory::submit! {
+    SourceDescription::new::<InternalMetricsConfig>("internal_metrics")
+}
+
+#[typetag::serde(name = "internal_metrics")]
+impl SourceConfig for InternalMetricsConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let fut = run(get_controller()?, out, shutdown).boxed().compat();
+        Ok(Box::new(fut))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "internal_metrics"
+    }
+}
+
+fn get_controller() -> crate::Result<Controller> {
+    crate::metrics::CONTROLLER
+        .get()
+        .cloned()
+        .ok_or("metrics system not initialized".into())
+}
+
+async fn run(
+    controller: Controller,
+    mut out: mpsc::Sender<Event>,
+    mut shutdown: ShutdownSignal,
+) -> Result<(), ()> {
+    let mut interval = interval(Duration::from_secs(2)).map(|_| ());
+
+    while let Some(()) = interval.next().await {
+        // Check for shutdown signal
+        if shutdown.poll().expect("polling shutdown").is_ready() {
+            break;
+        }
+
+        let metrics = capture_metrics(&controller);
+
+        let (sink, _) = out
+            .send_all(futures01::stream::iter_ok(metrics))
+            .compat()
+            .await
+            .expect("sending??");
```

Comment:
This can only fail if the topology task gets shutdown while we are sending, so I think we can ignore that here, instead of panicking?